flink-core/src/main/java/org/apache/flink/metrics/groups/AbstractMetricGroup.java

```diff
@@ -146,7 +147,16 @@ public Counter counter(int name) {
 
 	@Override
 	public Counter counter(String name) {
-		Counter counter = new Counter();
+		return counter(name, new SimpleCounter());
+	}
+	
+	@Override
+	public <C extends Counter> C counter(int name, C counter) {
```

Comment:
in other words, it allows the instantiation, registration and assignment to a variable or field in a single line without loosing the subclass type.